### DISCUSSION TITLE
PB2NC does not work for met/10.0.1 and metplus/4.0.0 for aqm daily max/ave verification

### DISCUSSION THREAD

**Discussion by Ho-ChunHuang-NOAA** [dtcenter/METplus#1484] [Statistical Computation] [answered]
Perry is transitioning the EMC AQM o3/pm25 verification package to me. Some
pb2nc functions worked before and do not work at present time.  Please help
us to resolve the problem.

script location
venus:/gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/scripts
script name : run_metplus_aqmax.sh

runtime_log output is
venus:/gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/scripts/met_help_run_metplus_aqmax.sh.log
Line: 202, 238, 275, 446, 483, 520
runtime_intermediated_log output are in
venus:/gpfs/dell2/ptmp/Ho-Chun.Huang/met_help_metplus_aqmax

We understand that we still need to work on the FCST_INPUT file.  But you
can ignore the error for " Could not find FCST_INPUT file" at this moment.

Perry also found the same script work for METplus 3.1 and MET 9.1.3.

**Comment by JohnHalleyGotway** [comment]
@Ho-ChunHuang-NOAA thanks for moving this topic over to METplus Discussions. I took a look at the METplus log files and see error messages stating that 0 observations were retained by PB2NC. I assume that's the issue that you're trying to debug.

FYI, I ran the following commands to replicate this behavior:
```
export METPLUS_LEVEL_CATEGORY=""; export METPLUS_LEVEL_RANGE_DICT=""; export METPLUS_MASK_DICT="mask = {grid = \"G212\";}"; export METPLUS_MESSAGE_TYPE=""; export METPLUS_MET_CONFIG_OVERRIDES=""; export METPLUS_OBS_BUFR_VAR="obs_bufr_var = [\"COPOPM\"];"; export METPLUS_OBS_WINDOW_DICT="obs_window = {beg = -43200;end = 43200;}"; export METPLUS_PB_REPORT_TYPE=""; export METPLUS_QUALITY_MARK_THRESH=""; export METPLUS_STATION_ID=""; export METPLUS_TIME_SUMMARY_DICT="time_summary = {flag = FALSE;beg = \"000000\";end = \"235959\";obs_var = [];type = [\"min\", \"max\", \"range\", \"mean\", \"stdev\", \"median\", \"p80\"];}"; export MET_TMP_DIR="/gpfs/dell2/emc/verification/noscrub/Ho-Chun.Huang/metplus_aq/tmp"; export OBS_BUFR_VAR_LIST="[\"COPOPM\"]"; export OBS_WINDOW_BEGIN="-43200"; export OBS_WINDOW_END="43200"; export PB2NC_GRID="G212"; export PB2NC_MESSAGE_TYPE="[]"; export PB2NC_POLY=""; export PB2NC_STATION_ID="[]"; export TIME_SUMMARY_BEG="\"000000\""; export TIME_SUMMARY_END="\"235959\""; export TIME_SUMMARY_FLAG="FALSE"; export TIME_SUMMARY_TYPES="[\"min\", \"max\", \"range\", \"mean\", \"stdev\", \"median\", \"p80\"]"; export TIME_SUMMARY_VAR_NAMES="[]";
export MET_TMP_DIR=~/
pb2nc -v 7 /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220302/aqm.t12z.anowpm.pb.tm024 ~/prepbufr.aqm.2022030100.nc /gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/parm/met_config/PB2NCConfig_pmmax -log ~/pb2nc.log
```

In the output I see many, many log messages like this:
```
DEBUG 4: Skipping time summary since the ratio of valid data 3/23 < 0.75 for ANOWPM, MEAN, 82800 seconds, 2FB15601, 20220301_040000, 19.53, -99.2, -9999, COPOPM, 0
```

You're running PB2NC in this case to compute 24-hourly maximum values from the observations. And I recall the challenge being that you want the maximum at 04Z for the previous 24 hours. Since each PREPBUFR file only contains 1 day's worth of obs, you actually need to provide it with 2 input files. To do so, you'd use the "-pbfile" command line option to pass the second input file.

It'd look something like this:
```
pb2nc -v 7 \
/gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220302/aqm.t12z.anowpm.pb.tm024 \
~/prepbufr.aqm.2022030100.nc \
/gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/parm/met_config/PB2NCConfig_pmmax \
-log ~/pb2nc.log \
-pbfile /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220301/aqm.t12z.anowpm.pb.tm024
```

Since you only gave it 24 hours of input, it only found data for 3 or 4 of the 24 hours it expects. And it didn't write any time summary outputs since it doesn't have enough valid input data.

I'll leave it to you and @PerryShafran-NOAA to iron out the details, but I believe that you'll need to use the "-pbfile" command line option to pass in a second input PrepBUFR file to compute a 04Z to 04Z maximum.

**Reply by PerryShafran-NOAA** [reply]
@JohnHalleyGotway I noticed some interesting things when comparing MET 9.1.3/METplus 3.1 vs MET 10.0/METplus 4.0.0, which I will be asking about at the upcoming METplus user's meeting.  Somehow the -pbfile command disappeared when moving to the later version.  No configuration files were changed between the two versions.  I think this will require some sort of deep dive as to why this is happening.

**Comment by georgemccabe** [comment, marked as answer]
It looks like there is a bug reading the file window variables for PB2NC wrapper. In METplus 3.1, PB2NC_FILE_WINDOW_[BEGIN/END] was supported but was inadvertently lost in a refactor.

Replacing PB2NC_FILE_WINDOW_BEGIN with OBS_PB2NC_FILE_WINDOW_BEGIN
and
replacing PB2NC_FILE_WINDOW_END with OBS_PB2NC_FILE_WINDOW_END
should fix this issue in 4.0.0.

I created GitHub issue https://github.com/dtcenter/METplus/issues/1486 to address this in METplus 4.1.0 so that PB2NC_FILE_WINDOW_[BEGIN/END] are still supported.

**Reply by PerryShafran-NOAA** [reply]
@georgemccabe @JohnHalleyGotway @Ho-ChunHuang-NOAA Something is still not right.  Let me compare some log output between the old run in 3.1 and the new run in 4.0.  

In the old run, here is the pb2nc command used, using some old versions of pb2nc:

 /gpfs/dell2/emc/verification/noscrub/Julie.Prestopnik/met/9.0/bin/pb2nc -v 4 /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220305/aqm.t12z.prepbufr.tm00 /gpfs/dell2/emc/verification/noscrub/Perry.Shafran/metplus_aq/aqmmax/aqmmax1/prepbufr.aqm.2022030500.nc /gpfs/dell2/emc/verification/save/Perry.Shafran/METplus-3.0-beta1/parm/met_config/PB2NCConfig_aqmmax -pbfile /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220306/aqm.t12z.prepbufr.tm00

And here is the summary output:

DEBUG 2: Total Messages processed               = 57456
DEBUG 2: Rejected based on message type         = 0
DEBUG 2: Rejected based on station id           = 0
DEBUG 2: Rejected based on valid time           = 0
DEBUG 2: Rejected based on masking grid         = 288
DEBUG 2: Rejected based on masking polygon      = 0
DEBUG 2: Rejected based on elevation            = 0
DEBUG 2: Rejected based on pb report type       = 0
DEBUG 2: Rejected based on input report type    = 0
DEBUG 2: Rejected based on instrument type      = 0
DEBUG 2: Rejected based on zero observations    = 7088
DEBUG 2: Total Messages retained                = 50080
DEBUG 2: Total observations retained or derived = 50080
DEBUG 1: Obs time between 20220306_010000 and 20220307_000000

and 

DEBUG 2: Total Messages processed               = 57456
DEBUG 2: Rejected based on message type         = 0
DEBUG 2: Rejected based on station id           = 0
DEBUG 2: Rejected based on valid time           = 0
DEBUG 2: Rejected based on masking grid         = 288
DEBUG 2: Rejected based on masking polygon      = 0
DEBUG 2: Rejected based on elevation            = 0
DEBUG 2: Rejected based on pb report type       = 0
DEBUG 2: Rejected based on input report type    = 0
DEBUG 2: Rejected based on instrument type      = 0
DEBUG 2: Rejected based on zero observations    = 6542
DEBUG 2: Total Messages retained                = 50626
DEBUG 2: Total observations retained or derived = 50626
DEBUG 1: Obs time between 20220305_010000 and 20220306_000000

This above was the good run using old versions:

Using the current version:

pb2nc -v 4 /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220305/aqm.t12z.prepbufr.tm00 /gpfs/dell2/emc/verification/noscrub/Perry.Shafran/metplus_aq/aqmmax/aqmmax1/prepbufr.aqm.2022030500.nc /gpfs/dell2/emc/verification/save/Perry.Shafran/METplus-4.0.0/parm/met_config/PB2NCConfig_aqmmax -pbfile /gpfs/dell1/nco/ops/com/hourly/prod/hourly.20220306/aqm.t12z.prepbufr.tm00

DEBUG 2: Total Messages processed               = 57456
DEBUG 2: Rejected based on message type         = 0
DEBUG 2: Rejected based on station id           = 0
DEBUG 2: Rejected based on valid time           = 55062
DEBUG 2: Rejected based on masking grid         = 12
DEBUG 2: Rejected based on masking polygon      = 0
DEBUG 2: Rejected based on elevation            = 0
DEBUG 2: Rejected based on pb report type       = 0
DEBUG 2: Rejected based on input report type    = 0
DEBUG 2: Rejected based on instrument type      = 0
DEBUG 2: Rejected based on zero observations    = 316
DEBUG 2: Total Messages retained                = 2066
DEBUG 2: Total observations retained or derived = 2066
DEBUG 1: Obs time between 20220306_010000 and 20220307_000000

DEBUG 2: Total Messages processed               = 57456
DEBUG 2: Rejected based on message type         = 0
DEBUG 2: Rejected based on station id           = 0
DEBUG 2: Rejected based on valid time           = 55062
DEBUG 2: Rejected based on masking grid         = 12
DEBUG 2: Rejected based on masking polygon      = 0
DEBUG 2: Rejected based on elevation            = 0
DEBUG 2: Rejected based on pb report type       = 0
DEBUG 2: Rejected based on input report type    = 0
DEBUG 2: Rejected based on instrument type      = 0
DEBUG 2: Rejected based on zero observations    = 297
DEBUG 2: Total Messages retained                = 2085
DEBUG 2: Total observations retained or derived = 2085
DEBUG 1: Obs time between 20220305_010000 and 20220306_000000

So the fix with OBS_PB2NC_FILE_WINDOW_BEGIN/END did bring in all the files needed, but for some reason it is now rejecting obs by valid time when the previous run did not.  (I didn't even realize the older version was a beta version, but it worked!)  So this needs a deeper dive to figure out what is wrong here.  Can you grab these files and take a look?

Thanks!

**Reply by PerryShafran-NOAA** [reply]
OK, I think I figured it out, perhaps.  I also needed to add in the following:

OBS_WINDOW_BEGIN = -86400
OBS_WINDOW_END = 86400

Somehow those were not needed in the older versions.  Once I added this in, all worked as expected.
